File: cpp/tests/packbin_tests.cpp
#include "packbin/packbin.hpp"

#include <chrono>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

namespace {

int failures = 0;

void expect(bool ok, char const* msg) {
  if (!ok) {
    std::cerr << "FAIL: " << msg << "\n";
    ++failures;
  }
}

std::vector<std::uint8_t> parse_hex(std::string hex) {
  while (!hex.empty() && (hex.back() == '\n' || hex.back() == '\r' || hex.back() == ' '))
    hex.pop_back();
  std::vector<std::uint8_t> out(hex.size() / 2);
  for (std::size_t i = 0; i < out.size(); ++i)
    out[i] = static_cast<std::uint8_t>(std::stoul(hex.substr(i * 2, 2), nullptr, 16));
  return out;
}

std::string find_golden() {
  char const* candidates[] = {
      "../fixtures/golden.hex",
      "fixtures/golden.hex",
      "../../fixtures/golden.hex",
  };
  for (auto const* path : candidates) {
    std::ifstream in(path);
    if (in) {
      std::stringstream buf;
      buf << in.rdbuf();
      return buf.str();
    }
  }
  return {};
}

auto position_scheme() {
  return packbin::scheme(0x40, {
      packbin::u16(0),
      packbin::i32(1),
      packbin::i32(2),
      packbin::u8(3),
      packbin::flags({packbin::u16(4), packbin::u8(5), packbin::i16(6)}),
  });
}

packbin::Values position_values() {
  packbin::Values v;
  v.emplace("0", packbin::Value{std::uint16_t{1}});
  v.emplace("1", packbin::Value{std::int32_t{500000000}});
  v.emplace("2", packbin::Value{std::int32_t{300000000}});
  v.emplace("3", packbin::Value{std::uint8_t{1}});
  return v;
}

constexpr char const* kGoldenHex = "4001000065cd1d00a3e1110100";

void ac1_position_pack() {
  auto bytes = packbin::BinaryPacker::pack(position_scheme(), position_values());
  auto hex = packbin::to_hex(bytes);
  expect(hex == kGoldenHex, "AC-1 hex");
  expect(packbin::mismatched_bytes(bytes, parse_hex(kGoldenHex)) == 0, "AC-1 mismatched");
  expect(bytes.size() == 13, "AC-1 length 13");
  auto again = packbin::BinaryPacker::pack(position_scheme(), position_values());
  expect(packbin::mismatched_bytes(bytes, again) == 0, "AC-1 packed twice");
}

void ac2_position_unpack() {
  packbin::Values got;
  auto result = packbin::BinaryPacker::unpack(
      parse_hex(kGoldenHex), position_scheme().on([&](packbin::Values const& row) { got = row; }));
  expect(result.ok, "AC-2 ok");
  expect(!packbin::present(got, "type"), "AC-2 no type");
  expect(std::get<std::uint16_t>(got.at("0").data) == 1, "AC-2 sid");
  expect(std::get<std::int32_t>(got.at("1").data) == 500000000, "AC-2 lat");
  expect(std::get<std::int32_t>(got.at("2").data) == 300000000, "AC-2 lon");
  expect(std::get<std::uint8_t>(got.at("3").data) == 1, "AC-2 profile");
  expect(packbin::motion_field_count(got) == 0, "AC-2 motion count 0");
  expect(!packbin::present(got, 4), "AC-2 no heading");
  expect(!packbin::present(got, 5), "AC-2 no speed");
  expect(!packbin::present(got, 6), "AC-2 no altitude");
}

void ac3_bytes_match_fixture() {
  auto fixture_text = find_golden();
  expect(!fixture_text.empty(), "AC-3 golden.hex found");
  auto fixture = parse_hex(fixture_text);
  auto bytes = packbin::BinaryPacker::pack(position_scheme(), position_values());
  expect(packbin::mismatched_bytes(bytes, fixture) == 0, "AC-3 mismatched 0");
}

void ac4_flags_and_stored_zero() {
  auto layout = packbin::scheme(1, {
      packbin::flags({packbin::u8(0), packbin::u8(1), packbin::u8(2), packbin::u8(3),
                      packbin::u8(4), packbin::u16(5)}),
  });

  auto clear = packbin::BinaryPacker::pack(layout, {});
  expect(clear.size() == 2, "AC-4 clear size 2");
  expect(clear[0] == 0x01 && clear[1] == 0x00, "AC-4 clear 0x01 0x00");

  packbin::Values set;
  set.emplace("5", packbin::Value{std::uint16_t{0x1234}});
  auto set_bytes = packbin::BinaryPacker::pack(layout, set);
  expect(set_bytes.size() == 4, "AC-4 0x20 size 4");
  expect(set_bytes[0] == 0x01 && set_bytes[1] == 0x20, "AC-4 flags 0x20");
  expect(set_bytes.size() - clear.size() == 2, "AC-4 adds 2 bytes");

  packbin::Values zero;
  zero.emplace("5", packbin::Value{std::uint16_t{0}});
  auto zero_bytes = packbin::BinaryPacker::pack(layout, zero);
  expect(zero_bytes.size() == 4, "AC-4 present 0 size");
  expect(zero_bytes[0] == 0x01 && zero_bytes[1] == 0x20, "AC-4 present 0 bit set");
  expect(zero_bytes[2] == 0x00 && zero_bytes[3] == 0x00, "AC-4 present 0 written");

  auto absent = packbin::BinaryPacker::pack(layout, {});
  expect(absent.size() == 2, "AC-4 absence size");
  expect(absent[0] == 0x01 && absent[1] == 0x00, "AC-4 absence no substitute");
  expect(absent.size() != zero_bytes.size(), "AC-4 absence != present 0");
}

void ac5_short_then_pack() {
  auto layout = packbin::scheme(1, {
      packbin::flags({packbin::u8(0), packbin::u8(1), packbin::u8(2), packbin::u8(3),
                      packbin::u8(4), packbin::u16(5)}),
  });
  bool ran = false;
  auto got = packbin::BinaryPacker::unpack(
      std::vector<std::uint8_t>{0x01, 0x20, 0x34},
      layout.on([&](packbin::Values const&) { ran = true; }));
  expect(!got.ok, "AC-5 not ok");
  expect(!ran, "AC-5 handler not run");
  expect(got.value_count() == 0, "AC-5 value count 0");
  expect(got.short_packet.has_value(), "AC-5 short packet");
  if (got.short_packet) {
    expect(got.short_packet->field == "5", "AC-5 field 5");
    expect(got.short_packet->needed == 2, "AC-5 needed 2");
    expect(got.short_packet->left == 1, "AC-5 left 1");
  }
  auto bytes = packbin::BinaryPacker::pack(position_scheme(), position_values());
  expect(packbin::to_hex(bytes) == kGoldenHex, "AC-5 pack after short");
}

void when_group_width() {
  auto layout = packbin::scheme(
      1, {packbin::u8(0),
          packbin::when(packbin::eq(0, packbin::Value{std::uint8_t{0}}), {packbin::u8(1)})});
  packbin::Values miss;
  miss.emplace("0", packbin::Value{std::uint8_t{1}});
  auto miss_bytes = packbin::BinaryPacker::pack(layout, miss);
  expect(miss_bytes.size() == 2, "when miss adds 0");

  packbin::Values hit;
  hit.emplace("0", packbin::Value{std::uint8_t{0}});
  hit.emplace("1", packbin::Value{std::uint8_t{9}});
  auto hit_bytes = packbin::BinaryPacker::pack(layout, hit);
  expect(hit_bytes.size() - miss_bytes.size() == 1, "when match adds group width");
}

void repeat_and_leftover() {
  auto layout = packbin::scheme(1, {packbin::repeat({packbin::u8(0), packbin::u8(1)})});
  packbin::Values ok_row;
  auto ok = packbin::BinaryPacker::unpack(
      std::vector<std::uint8_t>{1, 1, 2},
      layout.on([&](packbin::Values const& row) { ok_row = row; }));
  expect(ok.ok, "repeat ok");
  auto it = ok_row.find("0");
  expect(it != ok_row.end(), "repeat key");
  if (it != ok_row.end()) {
    auto const* list = std::get_if<packbin::Value::List>(&it->second.data);
    expect(list && *list && (*list)->items.size() == 1, "one group");
  }
  bool bad_ran = false;
  auto bad = packbin::BinaryPacker::unpack(
      std::vector<std::uint8_t>{1, 1, 2, 3},
      layout.on([&](packbin::Values const&) { bad_ran = true; }));
  expect(!bad.ok, "leftover not ok");
  expect(!bad_ran, "leftover handler not run");
  expect(bad.value_count() == 0, "leftover value count 0");
  expect(bad.short_packet.has_value(), "leftover short");
}

void trailing_byte() {
  bool ran = false;
  auto got = packbin::BinaryPacker::unpack(
      parse_hex(std::string(kGoldenHex) + "99"),
      position_scheme().on([&](packbin::Values const&) { ran = true; }));
  expect(!got.ok, "trailing not ok");
  expect(!ran, "trailing handler not run");
  expect(got.value_count() == 0, "trailing value count 0");
  expect(got.trailing && got.trailing->left == 1, "trailing left 1");
}

void assert_no_gpu() {
  std::ifstream in("/proc/self/maps");
  if (!in)
    return;
  std::stringstream buf;
  buf << in.rdbuf();
  auto blob = buf.str();
  for (auto const* bad : {"libcuda", "libnvidia", "libvulkan", "libopencl", "metal.framework"})
    expect(blob.find(bad) == std::string::npos, bad);
}

void nfr_round_trips() {
  auto layout = position_scheme();
  auto vals = position_values();
  auto start = std::chrono::steady_clock::now();
  packbin::Values last;
  bool ok = true;
  for (int i = 0; i < 100000; ++i) {
    auto bytes = packbin::BinaryPacker::pack(layout, vals);
    packbin::Values row;
    auto got = packbin::BinaryPacker::unpack(
        bytes, layout.on([&](packbin::Values const& v) { row = v; }));
    if (!got.ok) {
      ok = false;
      break;
    }
    last = std::move(row);
  }
  auto elapsed = std::chrono::steady_clock::now() - start;
  auto ms = std::chrono::duration<double, std::milli>(elapsed).count();
  expect(ok, "NFR unpack ok");
  expect(ok && std::get<std::int32_t>(last.at("1").data) == 500000000, "NFR lat");
  expect(ms <= 1000.0, "NFR <= 1s");
  assert_no_gpu();
  std::cerr << "nfr elapsed_ms " << ms << "\n";
}

}  // namespace

int run_kinds_tests();
int run_scheme_tests();
int run_field_id_binding_tests();
int run_borrowed_count_tests();

int main() {
  ac1_position_pack();
  ac2_position_unpack();
  ac3_bytes_match_fixture();
  ac4_flags_and_stored_zero();
  ac5_short_then_pack();
  when_group_width();
  repeat_and_leftover();
  trailing_byte();
  nfr_round_trips();
  failures += run_kinds_tests();
  failures += run_scheme_tests();
  failures += run_field_id_binding_tests();
  failures += run_borrowed_count_tests();
  if (failures != 0) {
    std::cerr << failures << " failure(s)\n";
    return 1;
  }
  std::cout << "all tests passed\n";
  return 0;
}

File: rust/tests/packbin_tests.rs
use packbin::{
    be, bytes, eq, f32, f64, flag_byte, flags, group, i16, i32, insert, mismatched_bytes,
    motion_field_count, pack_map, repeat, to_hex, u16, u8, unpack_map, when, MapScheme, ShortPacket,
    UnpackError, Value, Values,
};
use std::fs;
use std::time::Instant;

fn position_scheme() -> MapScheme {
    MapScheme::new(
        0x40,
        vec![
            u16("sid"),
            i32("lat"),
            i32("lon"),
            u8("profile"),
            flags(
                "motion",
                vec![u16("heading"), u8("speed"), i16("altitude")],
            ),
        ],
    )
}

fn position_values() -> Values {
    let mut v = Values::new();
    insert(&mut v, "sid", Some(Value::U16(1)));
    insert(&mut v, "lat", Some(Value::I32(500_000_000)));
    insert(&mut v, "lon", Some(Value::I32(300_000_000)));
    insert(&mut v, "profile", Some(Value::U8(1)));
    v
}

const GOLDEN_HEX: &str = "4001000065cd1d00a3e1110100";

fn parse_hex(hex: &str) -> Vec<u8> {
    (0..hex.len())
        .step_by(2)
        .map(|i| u8::from_str_radix(&hex[i..i + 2], 16).unwrap())
        .collect()
}

#[test]
fn ac1_position_pack() {
    let packed = pack_map(&position_scheme(), &position_values()).expect("pack");
    let hex = to_hex(&packed);
    assert_eq!(hex, GOLDEN_HEX);
    assert_eq!(mismatched_bytes(&packed, &parse_hex(GOLDEN_HEX)), 0);
    assert_eq!(packed.len(), 13);
    let again = pack_map(&position_scheme(), &position_values()).expect("pack again");
    assert_eq!(mismatched_bytes(&packed, &again), 0);
}

#[test]
fn ac2_position_unpack() {
    let got = unpack_map(&position_scheme(), &parse_hex(GOLDEN_HEX)).expect("unpack");
    assert!(!got.contains_key("type"));
    assert_eq!(got.get("sid"), Some(&Some(Value::U16(1))));
    assert_eq!(got.get("lat"), Some(&Some(Value::I32(500_000_000))));
    assert_eq!(got.get("lon"), Some(&Some(Value::I32(300_000_000))));
    assert_eq!(got.get("profile"), Some(&Some(Value::U8(1))));
    assert_eq!(motion_field_count(&got), 0);
    assert!(!got.contains_key("heading"));
    assert!(!got.contains_key("speed"));
    assert!(!got.contains_key("altitude"));
}

#[test]
fn ac3_bytes_match_fixture() {
    let path = concat!(env!("CARGO_MANIFEST_DIR"), "/../fixtures/golden.hex");
    let fixture = fs::read_to_string(path).expect("golden.hex").trim().to_string();
    let packed = pack_map(&position_scheme(), &position_values()).expect("pack");
    assert_eq!(to_hex(&packed), fixture);
    assert_eq!(mismatched_bytes(&packed, &parse_hex(&fixture)), 0);
}

#[test]
fn ac4_flags_and_stored_zero() {
    let scheme = MapScheme::new(
        0x40,
        vec![flags(
            "opts",
            vec![
                u8("b0"),
                u8("b1"),
                u8("b2"),
                u8("b3"),
                u8("b4"),
                u16("b5"),
            ],
        )],
    );

    let clear = Values::new();
    let clear_bytes = pack_map(&scheme, &clear).expect("pack clear");
    assert_eq!(clear_bytes.len(), 2);
    assert_eq!(clear_bytes[0], 0x40);
    assert_eq!(clear_bytes[1], 0x00);

    let mut set = Values::new();
    insert(&mut set, "b5", Some(Value::U16(0x1234)));
    let set_bytes = pack_map(&scheme, &set).expect("pack set");
    assert_eq!(set_bytes.len(), 4);
    assert_eq!(set_bytes[1], 0x20);
    assert_eq!(set_bytes.len() - clear_bytes.len(), 2);

    let mut zero = Values::new();
    insert(&mut zero, "b5", Some(Value::U16(0)));
    let zero_bytes = pack_map(&scheme, &zero).expect("pack zero");
    assert_eq!(zero_bytes.len(), 4);
    assert_eq!(zero_bytes[1], 0x20);
    assert_eq!(&zero_bytes[2..], &[0x00, 0x00]);

    let mut absent = Values::new();
    insert(&mut absent, "b5", None);
    let absent_bytes = pack_map(&scheme, &absent).expect("pack absent");
    assert_eq!(absent_bytes.len(), 2);
    assert_eq!(absent_bytes[1], 0x00);
    assert_eq!(absent_bytes.len(), clear_bytes.len());
    assert_ne!(zero_bytes.len(), absent_bytes.len());
}

#[test]
fn ac5_short_buffer_then_position_pack() {
    let scheme = MapScheme::new(
        1,
        vec![flags(
            "opts",
            vec![
                u8("b0"),
                u8("b1"),
                u8("b2"),
                u8("b3"),
                u8("b4"),
                u16("wide"),
            ],
        )],
    );
    let short = [0x01u8, 0x20, 0x34];
    let err = unpack_map(&scheme, &short).expect_err("short");
    match err {
        UnpackError::Short(ShortPacket {
            field,
            needed,
            left,
        }) => {
            assert_eq!(field, "wide");
            assert_eq!(needed, 2);
            assert_eq!(left, 1);
        }
        other => panic!("expected ShortPacket, got {:?}", other),
    }

    let packed = pack_map(&position_scheme(), &position_values()).expect("pack after short");
    assert_eq!(to_hex(&packed), GOLDEN_HEX);
}

#[test]
fn nfr_round_trips_under_one_second() {
    let scheme = position_scheme();
    let vals = position_values();
    let start = Instant::now();
    let mut last = Values::new();
    for _ in 0..100_000 {
        let packed = pack_map(&scheme, &vals).expect("pack");
        last = unpack_map(&scheme, &packed).expect("unpack");
    }
    let elapsed = start.elapsed();
    assert!(!last.contains_key("type"));
    assert_eq!(last.get("lat"), Some(&Some(Value::I32(500_000_000))));
    assert_no_gpu();
    assert!(
        elapsed.as_secs_f64() <= 1.0,
        "elapsed {:?} > 1s",
        elapsed
    );
    eprintln!("nfr elapsed_ms {:.1}", elapsed.as_secs_f64() * 1000.0);
}

#[test]
fn when_group_width() {
    let scheme = MapScheme::new(
        1,
        vec![
            u8("profile"),
            when(eq("profile", Value::U8(0)), vec![u8("shape")]),
        ],
    );
    let mut miss = Values::new();
    insert(&mut miss, "profile", Some(Value::U8(1)));
    insert(&mut miss, "shape", Some(Value::U8(9)));
    assert_eq!(pack_map(&scheme, &miss).unwrap().len(), 2);

    let mut hit = Values::new();
    insert(&mut hit, "profile", Some(Value::U8(0)));
    insert(&mut hit, "shape", Some(Value::U8(9)));
    assert_eq!(pack_map(&scheme, &hit).unwrap().len(), 3);
}

#[test]
fn repeat_groups_and_leftover() {
    let scheme = MapScheme::new(1, vec![u16("sid"), repeat(vec![i32("lat"), i32("lon")])]);
    let mut vals = Values::new();
    insert(&mut vals, "sid", Some(Value::U16(2)));
    let mut g1 = Values::new();
    insert(&mut g1, "lat", Some(Value::I32(10)));
    insert(&mut g1, "lon", Some(Value::I32(20)));
    insert(&mut vals, "__repeat__", Some(Value::Groups(vec![g1])));
    let packed = pack_map(&scheme, &vals).unwrap();
    let got = unpack_map(&scheme, &packed).unwrap();
    match got.get("__repeat__") {
        Some(Some(Value::Groups(g))) => assert_eq!(g.len(), 1),
        other => panic!("expected one group, got {:?}", other),
    }

    let mut leftover = packed;
    leftover.push(0xff);
    let err = unpack_map(&scheme, &leftover).expect_err("leftover");
    match err {
        UnpackError::Short(ShortPacket { field, .. }) => {
            assert!(field == "lat" || field == "lon");
        }
        other => panic!("expected short on leftover, got {:?}", other),
    }
}

#[test]
fn trailing_bytes_error() {
    let scheme = MapScheme::new(1, vec![u16("sid")]);
    let err = unpack_map(&scheme, &[1, 2, 3, 4]).expect_err("trailing");
    assert!(matches!(err, UnpackError::Trailing { left: 1 }));
}

#[test]
fn split_flag_byte_and_be() {
    let motion = flag_byte("motion");
    let scheme = MapScheme::new(
        1,
        vec![
            motion.byte(),
            when(eq("type", Value::U8(0)), vec![u8("shape")]),
            motion.bit(u16("heading")),
            motion.bit(u8("speed")),
        ],
    );
    let mut vals = Values::new();
    insert(&mut vals, "heading", Some(Value::U16(90)));
    let packed = pack_map(&scheme, &vals).unwrap();
    assert_eq!(packed[0], 0x01);
    assert_eq!(packed[1], 0x01);
    assert_eq!(packed.len(), 4);

    let be_scheme = MapScheme::new(1, vec![be(u16("sid"))]);
    let mut be_vals = Values::new();
    insert(&mut be_vals, "sid", Some(Value::U16(0x0102)));
    assert_eq!(pack_map(&be_scheme, &be_vals).unwrap(), vec![0x01, 0x01, 0x02]);

    let mut float_vals = Values::new();
    insert(&mut float_vals, "x", Some(Value::F32(1.0)));
    insert(&mut float_vals, "y", Some(Value::F64(2.0)));
    insert(&mut float_vals, "z", Some(Value::Bytes(vec![9, 8])));
    let float_scheme = MapScheme::new(1, vec![f32("x"), f64("y"), bytes("z", 2)]);
    assert_eq!(pack_map(&float_scheme, &float_vals).unwrap().len(), 15);
}

fn assert_no_gpu() {
    let Ok(text) = std::fs::read_to_string("/proc/self/maps") else {
        return;
    };
    let blob = text.to_ascii_lowercase();
    for bad in [
        "libcuda",
        "libnvidia",
        "libvulkan",
        "libopencl",
        "metal.framework",
    ] {
        assert!(!blob.contains(bad), "{bad}");
    }
}

#[test]
fn empty_group_flag() {
    let scheme = MapScheme::new(1, vec![flags("f", vec![group("mark", vec![])])]);
    let mut set = Values::new();
    insert(&mut set, "mark", Some(Value::U8(1)));
    assert_eq!(pack_map(&scheme, &set).unwrap(), vec![0x01, 0x01]);
    let clear = Values::new();
    assert_eq!(pack_map(&scheme, &clear).unwrap(), vec![0x01, 0x00]);
}

#[test]
fn flags_five_u8_then_u16() {
    let scheme = MapScheme::new(
        1,
        vec![flags(
            "f",
            vec![
                u8("a"),
                u8("b"),
                u8("c"),
                u8("d"),
                u8("e"),
                u16("b5"),
            ],
        )],
    );
    let mut a = Values::new();
    insert(&mut a, "a", Some(Value::U8(1)));
    assert_eq!(pack_map(&scheme, &a).unwrap().len(), 3);

    let mut b5 = Values::new();
    insert(&mut b5, "b5", Some(Value::U16(1)));
    let wide = pack_map(&scheme, &b5).unwrap();
    assert_eq!(wide[1], 0x20);
    let clear = pack_map(&scheme, &Values::new()).unwrap();
    assert_eq!(wide.len() - clear.len(), 2);
}

#[test]
fn session_group_pack_unpack() {
    let scheme = MapScheme::new(
        1,
        vec![flags(
            "f",
            vec![group("session", vec![u16("login"), packbin::u32("ts")])],
        )],
    );
    let mut vals = Values::new();
    insert(&mut vals, "login", Some(Value::U16(7)));
    insert(&mut vals, "ts", Some(Value::U32(1000)));
    let raw = pack_map(&scheme, &vals).unwrap();
    assert_eq!(to_hex(&raw[2..]), "0700e8030000");
    assert_eq!(raw.len() - 2, 6);

    let clear = pack_map(&scheme, &Values::new()).unwrap();
    assert_eq!(clear, vec![0x01, 0x00]);
    let got = unpack_map(&scheme, &clear).unwrap();
    assert!(!got.contains_key("login"));
    assert!(!got.contains_key("ts"));
}

#[test]
fn group_one_u8_zero() {
    let scheme = MapScheme::new(1, vec![flags("f", vec![group("g", vec![u8("b")])])]);
    let mut vals = Values::new();
    insert(&mut vals, "b", Some(Value::U8(0)));
    assert_eq!(pack_map(&scheme, &vals).unwrap(), vec![0x01, 0x01, 0x00]);
}

#[test]
fn session_group_short_read() {
    let scheme = MapScheme::new(
        1,
        vec![flags(
            "f",
            vec![group("session", vec![u16("login"), packbin::u32("ts")])],
        )],
    );
    let err = unpack_map(&scheme, &[0x01, 0x01, 0x07]).expect_err("short");
    match err {
        UnpackError::Short(ShortPacket {
            field,
            needed,
            left,
        }) => {
            assert_eq!(field, "login");
            assert_eq!(needed, 2);
            assert_eq!(left, 1);
        }
        other => panic!("expected ShortPacket, got {:?}", other),
    }
}

#[test]
fn map_scheme_type_number_range() {
    assert!(std::panic::catch_unwind(|| {
        let _ = MapScheme::new(256, vec![]);
    })
    .is_err());
}
